xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp

```diff
@@ -34,467 +39,1093 @@
+// static
+int CGUIDialogPVRTimerSettings::GetDateAsInt(const CDateTime &datetime)
+{
+  const CDateTime date(datetime.GetYear(), datetime.GetMonth(), datetime.GetDay(), 0, 0, 0);
+  time_t t(0);
+  date.GetAsTime(t);
+  return static_cast<int>(t);
+}
 
-  AddToggle(group, SETTING_TMR_ACTIVE, 19074, 0, m_bTimerActive);
-  AddEdit(group, SETTING_TMR_NAME, 19075, 0, tag->m_strTitle, false, false, 19097);
+// static
+void CGUIDialogPVRTimerSettings::SetDateFromInt(CDateTime &datetime, int date)
+{
+  const CDateTime newDate(static_cast<time_t>(date));
+  datetime.SetDateTime(
+    newDate.GetYear(), newDate.GetMonth(), newDate.GetDay(),
+    datetime.GetHour(), datetime.GetMinute(), datetime.GetSecond());
+}
 
-  if (tag->SupportsFolders())
-    AddEdit(group, SETTING_TMR_DIR, 19076, 0, tag->m_strDirectory, true, false, 19104);
+// static
+void CGUIDialogPVRTimerSettings::SetTimeFromSystemTime(CDateTime &datetime, const SYSTEMTIME &time)
+{
+  const CDateTime newTime(time);
+  datetime.SetDateTime(
+    datetime.GetYear(), datetime.GetMonth(), datetime.GetDay(),
+    newTime.GetHour(), newTime.GetMinute(), newTime.GetSecond());
+}
 
-  /// Channel names
+// static
+CDateTime CGUIDialogPVRTimerSettings::InitializeDateTime(const CDateTime &datetime)
+{
+  const CDateTime now(CDateTime::GetCurrentDateTime());
+  CDateTime dt(datetime);
+  if (now > dt)
+    dt = now;
+
+  return dt;
+}
+
+void CGUIDialogPVRTimerSettings::InitializeTypesList()
+{
+  m_typeEntries.clear();
+
+  int idx = 0;
+  const std::vector<CPVRTimerTypePtr> types(CPVRTimerType::GetAllTypes());
+  for (const auto &type : types)
   {
-    AddChannelNames(group, tag->m_bIsRadio);
+    if (m_bIsNewTimer)
+    {
+      // Type definition prohibits created of new instances.
+      // But the dialog can act as a viewer for these types. (m_bIsNewTimer is false in this case)
+      if (type->ForbidsNewInstances())
+        continue;
+
+      // Read-only timers cannot be created using this dialog.
+      // But the dialog can act as a viewer for read-only types (m_bIsNewTimer is false in this case)
+      if (type->IsReadOnly())
+        continue;
+
+      // For new timers, skip one time epg-based types. Those cannot be created using this dialog (yet).
+      if (type->IsOnetimeEpgBased())
+        continue;
+    }
+
+    m_typeEntries.insert(std::make_pair(idx++, type));
   }
+}
 
-  /// Day
-  {
-    // get diffence of timer in days between today and timer start date
-    tm time_cur; CDateTime::GetCurrentDateTime().GetAsTm(time_cur);
-    tm time_tmr; tag->StartAsLocalTime().GetAsTm(time_tmr);
+void CGUIDialogPVRTimerSettings::InitializeChannelsList()
+{
+  m_channelEntries.clear();
+
+  CFileItemList channelsList;
+  g_PVRChannelGroups->GetGroupAll(m_bIsRadio)->GetMembers(channelsList);
 
-    m_tmp_day += time_tmr.tm_yday - time_cur.tm_yday;
-    if (time_tmr.tm_yday - time_cur.tm_yday < 0)
-      m_tmp_day += 365;
+  for (int i = 0; i < channelsList.Size(); ++i)
```

Comment:
not related to this pr really, but just wondering cause i looked up the method, thinking this should be `size_t` rather than `int`, but why do we just do a cast to `int` in `int CFileItemList::Size() const`?